Documentation/botr/shared-generics.md

```diff
@@ -0,0 +1,185 @@
+``` asm
+    mov rcx, type handle of List<string> or List<object>
+    call ToString()
+    ret
+```
+
+With shared generics, the canonical code will not have any hard-coded versions of the type handle of List<T>, but instead looks up the exact type handle either through a call to a runtime helper API, or by loading it up from the *generic dictionary* of the instantiation of Func<T> that is executing. The code would look more like the following:
+``` asm
+    mov rcx, generic context                                                // MethodDesc of Func<string> or Func<object>
+    mov rcx, [rcx + offset of InstantiatedMethodDesc::m_pPerInstInfo]       // This is the generic dictionary
+    mov rcx, [rcx + dictionary slot containing type handle of List<T>]
+    call ToString()
+    ret
+```
+
+The generic context in this example is the InstantiatedMethodDesc of `Func<object>` or `Func<string>`. The generic dictionary is a data structure used by shared generic code to fetch instantiation-specific information. It is basically an array where the entries are instantiation-specific type handles, method handles, field handles, method entry points, etc... The "PerInstInfo" fields on MethodTable and InstantiatedMethodDesc structures point at the generic dictionary structure for a generic type and method respectively.
+
+In this example, the generic dictionary for Func<object> will contain a slot with the type handle for type List<object>, and the generic dictionary for Func<string> will contain a slot with the type handle for type List<string>.
+
+This feature is currently only supported for instantiations over reference types because they all have the same size/properties/layout/etc... For instantiations over primitive types or value types, the runtime will generate separate code bodies for each instantiation.
+
+
+# Layouts and Algorithms
+
+### Dictionaries Pointers on Types and Methods
+
+The dictionary used by any given generic method is pointed at by the `m_pPerInstInfo` field on the `InstantiatedMethodDesc` structure of that method. It's a direct pointer to the contents of the generic dictionary data.
+
+On generic types, there's an extra level of indirection: the 'm_pPerInstInfo' field on the `MethodTable` structure is a pointer to a table of dictionaries, and each entry in that table is a pointer to the actual generic dictionary data. This is because types have inheritance, and derived generic types inherit the dictionary pointers of their base types. 
+
+Here's an example:
+```c#
+class BaseClass<T> { }
+
+class DerivedClass<U> : BaseClass<U> { }
+
+class AnotherDerivedClass : DerivedClass<string> { }
+```
+
+The MethodTables of each of these types will look like the following:
+
+| **BaseClass[T]'s MethodTable** |
+|--------------------------|
```

Comment:
It would be nice to pad these as necessary, so that tables are readable as raw text.